django/thunderstore/community/models/community.py

```diff
@@ -206,3 +199,140 @@ def valid_review_statuses(self):
+    package_count = models.PositiveIntegerField(default=0)
+
+    def __str__(self):
+        return str(self.as_class())
+
+    @classmethod
+    def from_dataclass(cls, fields: AggregatedFields) -> "CommunityAggregatedFields":
+        return cls(
+            download_count=fields.download_count,
+            package_count=fields.package_count,
+        )
+
+    @classmethod
+    def get_empty(cls) -> AggregatedFields:
+        return AggregatedFields()
+
+    def as_class(self) -> AggregatedFields:
+        """
+        Return fields in a format supporting attribute access.
+        """
+        return AggregatedFields(self.download_count, self.package_count)
+
+    @classmethod
+    @transaction.atomic
+    def update_for_communities(
+        cls,
+        communities: QuerySet[Community],
+        batch_size: Optional[int] = None,
+    ) -> None:
+        """
+        Updates field values for given Communities.
+
+        Creates the CommunityAggregatedFields object and updates the
+        reference on Community object if none existed previously.
+        """
+        aggregated = cls._aggregate(communities)
+        communities_with_af = [c for c in communities if c.aggregated_fields]
+        communities_without_af = [c for c in communities if not c.aggregated_fields]
+
+        # PrimaryKey and ForeignKey restrictions requires extra steps for
+        # communitites that doesn't have aggregated_fields reference yet.
+        created_afs = cls.objects.bulk_create(
+            [cls.from_dataclass(aggregated[c.id]) for c in communities_without_af],
+            batch_size=batch_size,
+        )
+
+        for community, aggregated_fields in zip(communities_without_af, created_afs):
+            community.aggregated_fields = aggregated_fields
+
+        Community.objects.bulk_update(
+            communities_without_af,
+            ["aggregated_fields"],
+            batch_size,
+        )
+
+        # Update field values for pre-existing CommunityAggregatedFields.
+        updated = [
+            c.aggregated_fields._update_fields(aggregated[c.id])
+            for c in communities_with_af
+        ]
+        fields = ["download_count", "package_count"]
+        cls.objects.bulk_update(updated, fields, batch_size=batch_size)
+
+    @classmethod
+    def _aggregate(
+        cls,
+        communities: QuerySet[Community],
+    ) -> DefaultDict[int, AggregatedFields]:
+        """
+        Precalculates the values for download_count and package_count.
+
+        Helper method for update_for_communities.
+        """
+        communities = communities.prefetch_related(
+            "package_listings__package__versions",
+        )
+        results: DefaultDict[int, AggregatedFields] = defaultdict(
+            lambda: AggregatedFields(),
+        )
+
+        # Manually loop through the results and ignore deactiveted items
+        # etc. to avoid the extra database hits caused by using the
+        # existing convenience methods.
```

Comment:
This also sounds like a good way to not have changes to the visibility rules appropriately reflected in the counts here. I'd say code simplicity is much preferred over any performance in this case given that it doesn't really matter what the runtime of this task is.

Reply:
Implementation is now based on existing queryset helpers. E.g. downloads for deactivated package versions are included in the results, unless all package versions of a package are deactivated, in which case the package is ignored completely.